src/types/bytes.rs

```diff
@@ -102,16 +128,9 @@ impl Py<PyBytes> {
     /// immutable, the result may be used for as long as the reference to
     /// `self` is held, including when the GIL is released.
     pub fn as_bytes<'a>(&'a self, _py: Python<'_>) -> &'a [u8] {
-        // py is required here because `PyBytes_AsString` and `PyBytes_Size`
-        // can both technically raise exceptions which require the GIL to be
-        // held. The only circumstance in which they raise is if the value
-        // isn't really a `PyBytes`, but better safe than sorry.
-        unsafe {
-            let buffer = ffi::PyBytes_AsString(self.as_ptr()) as *const u8;
-            let length = ffi::PyBytes_Size(self.as_ptr()) as usize;
-            debug_assert!(!buffer.is_null());
-            std::slice::from_raw_parts(buffer, length)
-        }
+        // Safety: `Py2Borrowed` has the same layout as `Py`
+        let borrowed: Py2Borrowed<'a, '_, PyBytes> = unsafe { std::mem::transmute_copy(self) };
+        borrowed.as_bytes()
```

Comment:
Ok, so this didn't work out as nicely as I hoped. I _thought_ I could have the following implementation:

```rust
    pub fn as_bytes<'a>(&'a self, py: Python<'_>) -> &'a [u8] {
        Py2Borrowed::from(self.attach(py)).as_bytes()
    }
```

... but the compiler wants to have an explicit lifetime `Python<'py>` with a constraint `'py: 'a`. I feel like that might be a technical weakness in the current borrow checker which could be overcome with time, but it could be that I've misunderstood some fine detail about lifetimes. I'll try to ask language folks about this on Zulip.

Reply:
Ok, so I asked in https://users.rust-lang.org/t/is-this-a-limitation-of-the-current-borrow-checker/103979/3 and I think I have some ideas now, so I'll come back and push a solution later today.

Reply:
Ok, this now works; I had to adjust the definition of the ``Py2Borrowed` struct slightly, no unsafe needed any more :)